### PR TITLE
Updated to build against current interfaces.

### PR DESCRIPTION
**Pull request by joecrayne**
I've made a lot of edits to get the code to build against current library interfaces.  It now uses a newer AESNI-aware interface to c code as the Haskell implementation was removed from hs-cryptocipher.  hs-cryptocipher also removed their BlockCipher instance for Bluefish without providing sufficient access to internals.  So to build, you will need my fork (https://github.com/joecrayne/hs-crypto-cipher/commit/5858a64a8e4315ab0bb98b4805520ae11e9086c6) which I've requested be pulled upstream (https://github.com/vincenthz/hs-crypto-cipher/pull/27)

### CODE REVIEW

**Comment by singpolyma** [comment]
Apparently the build is failing on this?

I need to sit down and think about the CPRG problem.  I do not want to use CPRG in the public API, and need to make sure that all internal uses are checked and secure (since CPRG will repeat bytes on you if you aren't careful to ask it how many are left, which can lead to security problems).

**Comment by joecrayne** [comment]
Isn't CPRG just a class interface that exports the same underlying implementations as the one it replaced?  If you know a link where I can read about the repeating bytes issue, I'm interested.

**Comment by singpolyma** [comment]
`CPRG` is the result of a disagreement between the author of `crypto-api` (where CryptoRandomGen, the safe class, lives) and the author of `cryptocipher`/`crypto-pubkey`.  They exist in parallel.  More discussion can be found at https://github.com/vincenthz/hs-crypto-pubkey/issues/1 https://github.com/TomMD/crypto-api/issues/13 https://github.com/vincenthz/hs-crypto-pubkey/issues/5 https://github.com/vincenthz/hs-crypto-pubkey/issues/4

Key quote is:

> The problem wtih crypto-api random's api is that it is .. well too accurate.

I believe that being accurate is key to security, and so do not want to expose use of `CPRG` to my public API, lest a user accidentally use it wrong.  I want the OpenPGP libraries to provide safety to users so that they can get on with the important parts of their app.

**Comment by singpolyma** [comment]
Ah, I see.  The tests are failing because https://github.com/vincenthz/hs-crypto-cipher/pull/27 hasn't gone in yet.

**Comment by joecrayne** [comment]
Would it be acceptable to hide the random number generator from the API and just internally always use System.Random so you don't have to worry about reseeding?  I don't know why I'd want to use some other generator to create crypto keys.

**Comment by singpolyma** [comment]
I've thought about that, but that introduces `IO` to the API, and does make it less flexible.  I think exposing `CryptoRandomGen` and then wrapping in a newtype that makes it also `CPRG` and then carefully checking that there is enough entropy before calling any APIs is the right way to do it.

**Comment by singpolyma** [comment]
https://github.com/vincenthz/hs-cipher-aes/issues/6 is also an issue, since this package wants to use `crypto-api` wherever possible.

**Comment by singpolyma** [comment]
I have released a version of this package with a more restrictive version requirement, so that it can at least be installed.

File: Data/OpenPGP/CryptoAPI/AES.hs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+instance AESMode M192 where aesKeyLength _ = 192
+instance AESMode M256 where aesKeyLength _ = 256
+
+instance AESMode mode => Serialize (AES mode) where
+    put (AES k) = putByteString . keyOfCtx $ k
+    get = do
+        bs <- getBytes (aesKeyLength (undefined::mode) `div` 8) 
+        return . AES . initKey $ bs
+            
+instance AESMode mode => BlockCipher (AES mode) where
```
Comment (singpolyma) [inline Data/OpenPGP/CryptoAPI/AES.hs]: This file is just about getting a `BlockCipher` instance back for AES?  Should probably get cleaned up a bit and then put in it's own package.  If Vincent refuses, I'll just make a `cryptocipher-cryptoapi` package with wrappers for all the things.  The Blowfish could live there as well.  I'll wait a little to see what Vincent does, but yeah, this seems like the right solution.

**Comment by joecrayne** [comment]
I think you should consider providing a convenience API that does add IO and uses /dev/random or whatever for good entropy, as I think most of your users will not want to be bothered about the random number generator.  The flexible interface can be moved off to a .Pure module.  That will also insulate your users in case there continues to be upheaval in the random APIs.

**Comment by singpolyma** [comment]
Since this particular wrapper is dedicated to `crypto-api`, it will always use that random interface (no upheaval) -- however, I will consider adding an IO-based one as well, since that's not an unreasonable thing to want.

**Comment by singpolyma** [comment]
https://github.com/singpolyma/OpenPGP-Crypto for example wraps a different set of stuff.

**Comment by joecrayne** [comment]
It was decryptSecretKey that landed me on here.  Is that available elsewhere?

**Comment by singpolyma** [comment]
This library is the wrapper I most actively maintain, so `decryptSecretKey` is only here for now.

I think once https://github.com/vincenthz/hs-cipher-aes/issues/6 goes one way or the other, I should be able to make more progress on this.

I have pushed a new version that works with the latest `crypto-api` at least.

**Comment by joecrayne** [comment]
closed because github wouldn't let me make another pull request.
I notice also github seems to be showing my current "master" branch in the diff.
So for reference: the original content is a single commit and was moved to my "pull4" branch.
1bc5c1cae4398bb6dd460dd0a83ed5b4a7eafac0

**Comment by singpolyma** [comment]
#6 now tracks progress on this

**Pull request closed without merging**